### PR TITLE
fix: Upgrade code detection

### PR DESCRIPTION
**Pull request by denizkirbiyik**
I replaced the system used for code detection with a different system that uses code-specific regex patterns/expressions.

The previous system was prone to false positives.

### CODE REVIEW

File: cogs/detect_code.py [hunk 1]
```diff
@@ -71,15 +74,25 @@ def likely_contains_code(text: str) -> bool:
...
+            r'"""',
+            r";(?=\s*(//|#).+|$)",
+            r"^[\w]+\.[\w]+(\.[\w]+)*$",
+            r"\b\w+(?:_\w+)*\b(?:->\b\w+(?:_\w+)*\b)+",
+            r"^\/\*",
+            r".*\*/$",
+            r"[A-Za-z]+(?:[a-z][A-Z]|[A-Z][a-z])[A-Za-z]*",
+            r"^\w+(?:_\w+)+$",
+            r"\b[a-zA-Z0-9]+(?:-[a-zA-Z0-9]+)+\b",
+        ]
```
Comment (GabeMillikan) [inline cogs/detect_code.py:91]: Some suggestions to simplify these patterns, in the context of how they're used:
1.`\b\w+\s*\(\s*.*?\s*\)\s*` is equivalent to `\w+\s*\(.*\)` 
2. `(#.*$)` is equivalent to `#` - any line containing a hashtag
3. `;(?=\s*(//|#).+|$)` is equivalent to `;\s*(//|#)`
4. `^[\w]+\.[\w]+(\.[\w]+)*$` is equivalent to `^\w+(\.\w+)+$`
5. `\b\w+(?:_\w+)*\b(?:->\b\w+(?:_\w+)*\b)+` is equivalent to `\w+(->\w+)+`
6. `.*\*/$` is equivalent to `\*/$`
7. `[A-Za-z]+(?:[a-z][A-Z]|[A-Z][a-z])[A-Za-z]*` is equivalent to `[A-Za-z]+([a-z][A-Z]|[A-Z][a-z])` and I would say that `[a-z][A-Z]` would be good enough on its own

File: cogs/detect_code.py [hunk 2]
```diff
@@ -71,15 +74,25 @@ def likely_contains_code(text: str) -> bool:
...
+        combined_pattern = "|".join(patterns)
```
Comment (GabeMillikan) [inline cogs/detect_code.py:92]: Combining the patterns like this scares me - it works, but in the future people may try to add new patterns which _are not_ compatible with this pipe join. I would just leave them separate, and later do `if any(re.search(line, pattern) for pattern in patterns)`

**Review by GabeMillikan — CHANGES_REQUESTED** [review]
Thanks so much for the contribution! This feature has needed some TLC for a while...

Can you add some examples of what is/isn't detected as code to your PR description please?

I have noticed that Pax's tips are usually sent erroneously when somebody has 2-4 lines of text and they have commas or parentheses or whatever. I think it would be worth trying to restrict Pax to only consider code if it's at least 5-6 lines of text; nobody ever sends just two lines of code.

File: cogs/detect_code.py (hunk 1 above)
Comment (denizkirbiyik) [reply cogs/detect_code.py:91]: 3 is an incorrect fix as it does not account for there being no comment after the semicolon. Everything else seems reasonable.

File: cogs/detect_code.py (hunk 2 above)
Comment (denizkirbiyik) [reply cogs/detect_code.py:92]: Sure, I'll do that.

File: cogs/detect_code.py (hunk 1 above)
Comment (GabeMillikan) [reply cogs/detect_code.py:91]: I see 👍

**Comment by GabeMillikan** [comment]
Thank you, can you share some of your tests here? It would be nice to paste a couple of examples in your PR description that demonstrate:
- code that _will_ be detected, which wasn't previously detected
- english that _won't_ be detected, that was previously detected incorrectly

**Comment by denizkirbiyik** [comment]
Things that was previously falsely detected as code, but were correctly detected now:

`Amber ( Inventory Personnel )
Search function for PRODUCT  [(search product you need) and (search Expired product)]
Replace or ament BALANCE QUANTITY, QUANTITY  (for Old Product)
Delete expired product ( delete only quantity and balance quantity )
[Quantity must be 0 for expiry products, 0 to represent delete]
Exit ( exit inventory )

This is the notes I've made`

`PA seems prone to false positives, for example:
Lines ending with (semi)colons
Text in parenthesies (be careful with them)`

`(wow)
(no wonder there are so many false positives)`

`My process for this question:

-16+8+4+1 = -3 (Exponent)

Next I moved the decimal point 3 places to the right

Answer = 0101.010 = 5.25 in denary


The correct answer is: 0.08203125 (However, I do not understand how to get to this answer) 
`

`How to optimize the following Boolean expression using a map:
X'Z' + YZ' + XYZ`

`i will do it!!

(i am lying)`

`Hi, so i obtained the values:
27 13 34 85 59 91 50 38 as my step 1, in which 34 is sorted. However, i only have 2 values 27 and 13. How do i perform another middle of 3?`

`Theorem 4.2: All recursive algorithms must have a base case.
  Proof: See Theorem 4.2.`

`1: searches for all occurrences of 1 (1)
2: searches for all occurrences of 2 (2)
2: searches for all occurrences of 2 (2)
4: searches for all occurrences of 4 (2)
4: searches for all occurrences of 4 (2)
2 + 2 + 2 + 2 + 1 = 9`

`would you break it like:

$2(3^k) + 3^k - 6 > 4k + 4$`

`so from what i'm getting is:

learn the math (discrete, stats & probability, algebra, calc (for fun))`

`what if you do this instead:
Inside of recv you read the new frame. If you're not processing a frame you start transforming it async without awaiting (and set the boolean to true). Then return the last processed frame. The job of transform_frame is to set last_frame inside of itself (so it doesn't return anything). Or, if there is some kind of similar thing to .then as in JS, do that instead`

`if you want to develop native windows apps then i guess the following might be a decent option
clang/llvm as compiler (works just like gcc)
clangd extension for c language support in vscode
ninja for building (replaces make)
cmake + cmake extension in vscode for c project management`

**Review by danparizher — APPROVED** [review]

**Review by sebastiaan-daniels — APPROVED** [review]

**Comment by GabeMillikan** [comment]
Thanks again for your effort here! Unfortunately we've decided to simply [delete the feature altogether](https://github.com/danparizher/Pax-Academia/pull/283), because it is generally just annoying and not helpful. This PR _would have_ fixed a lot false-positives where non-code was detected as code, but it does not meaningfully improve the correct-positives where code is detected as code. Doing so is a difficult challenge, and is better solved by language-specific rules in the `./code_detection/` directory.

**Pull request closed without merging**